dhis2/dhis2-core
# feat: add CSRF token feature

## Summary
Introduces config option to enable CSRF tokens in Spring security. Historically this feature has been disabled due to the high cost of implementing it, caused by the diverse ecosystem of DHIS2 apps. Since 2.42 all apps are SPA apps and all bundled apps are using a common library for requests, this feature now seems more realistic to be implemented.
This PR serves as a development preview of the feature, and is disabled by default. The idea is that it can be enabled when app developers start integrating/testing it into their apps.

The PR implements the reference code for the SPA version of CSRF tokens, explained here: [https://docs.spring.io/spring-security/reference/servlet/exploits/csrf.html#csrf-integration-javascript-spa](https://docs.spring.io/spring-security/reference/servlet/exploits/csrf.html#csrf-integration-javascript-spa)

The server will set a HttpOnly=false cookie called XSRF-TOKEN, which the client needs to extract and put into a header called `X-XSRF-TOKEN` for all POST, PUT, PATCH and DELETE requests.

### Automatic tests
N/A yet.

### Manual testing
1. Enable the feature in the dhis.conf with: `http.security.csrf.enabled = true`
2. Install ModHeader Chrome extension or similar to "fake" the CSRF token header.
3. Do any requests to the server and fetch the XSRF-TOKEN cookie value that should be set.
4. Add a header called: `X-XSRF-TOKEN`, in the ModHeader extension, use the value fetched in step 3.
5. Observe you don't get a 503 forbidden when trying to do a POST, PUT, PATCH and DELETE.
